pkg/steps/template.go

```diff
@@ -421,24 +421,77 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
-	timeout := 300
-	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
-		}
+func waitForTeardownToComplete(podClient coreclientset.PodInterface, name string, uid types.UID) {
+	timeout := 5 * time.Minute
+
+	log.Printf("Waiting for pod %s to complete teardown ...", name)
+	wait.Poll(10*time.Second, timeout, func() (done bool, err error) {
+		pod, err := checkPodExistsAndValid(podClient, name, uid)
 		if err != nil {
-			return fmt.Errorf("could not retrieve deleting pod: %v", err)
+			// Pod was destroyed
+			return true, nil
 		}
-		if pod.UID != uid {
-			return nil
+
+		for _, status := range append(append([]coreapi.ContainerStatus{}, pod.Status.InitContainerStatuses...), pod.Status.ContainerStatuses...) {
+			if status.Name == "teardown" && status.State.Terminated != nil {
+				// Teardown container finished
+				return true, nil
+			}
 		}
-		log.Printf("Waiting for pod %s to be deleted ... (%ds/%d)", name, i, timeout)
-		time.Sleep(2 * time.Second)
+		return false, nil
+	})
+}
+
+// Check that pod with this name exists and has the same UID
+func checkPodExistsAndValid(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, errors.NewGone("pod was removed")
 	}
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+	if pod.UID != uid {
+		return nil, errors.NewBadRequest("mismatching UID")
+	}
+	return pod, nil
+}
 
-	return fmt.Errorf("waited for pod %s deletion for %ds, was not deleted", name, timeout)
+func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+	timeout := 5 * time.Minute
+	pod, err := checkPodExistsAndValid(podClient, name, uid)
+	if err != nil {
+		// Pod with this name and UID is gone, so its successfully deleted
```

Comment:
or, the `fmt.Errorf("could not retrieve deleting pod: %v", err)` on L452 happened, which is likely some infra error and should bubble up. 

I think it may be fine to assume that either the pod does not exist or the infra error will bring down the run a bit later, but this assumption would need to be documented here.

Reply:
Passing infra errors in d2eb3c now, PTAL